hermes-client/src/main/java/pl/allegro/tech/hermes/client/HermesClient.java

```diff
@@ -93,18 +108,24 @@ private String createUri(URI uri) {
         return publishWithRetries(message);
     }
 
+    public boolean addMessageDeliveryListener(MessageDeliveryListener listener) {
+        return messageDeliveryListeners.add(listener);
+    }
+
     private CompletableFuture<HermesResponse> publishWithRetries(HermesMessage message) {
         currentlySending.incrementAndGet();
         return Failsafe.with(retryPolicy)
                 .with(scheduler)
-                .future(() -> sendOnce(message))
-                .whenComplete((response, ex) -> currentlySending.decrementAndGet());
+                .onComplete((e) -> currentlySending.decrementAndGet())
+                .getStageAsync(() -> sendOnce(message));
     }
 
     private CompletableFuture<HermesResponse> sendOnce(HermesMessage message) {
-        return sender
-                .send(URI.create(uri + message.getTopic()), message)
-                .exceptionally(e -> HermesResponseBuilder.hermesFailureResponse(e, message));
+        long startTime = System.nanoTime();
+
+        return sender.send(URI.create(uri + message.getTopic()), message)
+                .exceptionally(e -> HermesResponseBuilder.hermesFailureResponse(e, message))
+                .whenComplete((resp, cause) -> messageDeliveryListeners.forEach(l -> l.onSend(resp, System.nanoTime() - startTime)));
```

Comment:
I wonder if it's a problem that every listener will get a little bit different time.

Reply:
Mybe we could make a single variable and pass it to each one.

Reply:
btw, shouldn't we pass the difference of current time and start time (latency)?